plan/aggregation_push_down.go

```diff
@@ -0,0 +1,246 @@
+		leftGbyCols = addGbyCol(leftGbyCols, eqFunc.Args[0].(*expression.Column))
+		rightGbyCols = addGbyCol(rightGbyCols, eqFunc.Args[1].(*expression.Column))
+	}
+	for _, leftCond := range join.LeftConditions {
+		var cols []*expression.Column
+		cols, _ = extractColumn(leftCond, cols, nil)
+		leftGbyCols = addGbyCol(leftGbyCols, cols...)
+	}
+	for _, rightCond := range join.RightConditions {
+		var cols []*expression.Column
+		cols, _ = extractColumn(rightCond, cols, nil)
+		rightGbyCols = addGbyCol(rightGbyCols, cols...)
+	}
+	for _, otherCond := range join.OtherConditions {
+		var cols []*expression.Column
+		cols, _ = extractColumn(otherCond, cols, nil)
+		for _, col := range cols {
+			if leftChild.GetSchema().GetIndex(col) != -1 {
+				leftGbyCols = addGbyCol(leftGbyCols, col)
+			} else {
+				rightGbyCols = addGbyCol(rightGbyCols, col)
+			}
+		}
+	}
+	return
+}
+
+func checkValidAgg(agg *Aggregation, join *Join) (valid bool,
+	leftAggFuncs, rightAggFuncs []expression.AggregationFunction,
+	leftGbyCols, rightGbyCols []*expression.Column) {
+	valid, leftAggFuncs, rightAggFuncs = collectAggFuncs(agg, join)
+	if !valid {
+		return
+	}
+	leftGbyCols, rightGbyCols = collectGbyCols(agg, join)
+	return
+}
+
+func addGbyCol(gbyCols []*expression.Column, cols ...*expression.Column) []*expression.Column {
+	for _, c := range cols {
+		duplicate := false
+		for _, gbyCol := range gbyCols {
+			if c.Equal(gbyCol) {
+				duplicate = true
+				break
+			}
+		}
+		if !duplicate {
+			gbyCols = append(gbyCols, c)
+		}
+	}
+	return gbyCols
+}
+
+func checkValidJoin(join *Join) bool {
+	// TODO: Support outer join.
+	return join.JoinType == InnerJoin
+}
+
+func decompose(aggFunc expression.AggregationFunction, schema expression.Schema, id string) ([]expression.AggregationFunction, expression.Schema) {
+	var result []expression.AggregationFunction
+	switch aggFunc.GetName() {
+	case ast.AggFuncAvg:
+		count := expression.NewAggFunction(ast.AggFuncCount, aggFunc.GetArgs(), false)
+		sum := expression.NewAggFunction(ast.AggFuncSum, aggFunc.GetArgs(), false)
+		result = []expression.AggregationFunction{count, sum}
+	default:
+		result = []expression.AggregationFunction{aggFunc.Clone()}
+	}
+	for _, aggFunc := range result {
+		schema = append(schema, &expression.Column{
+			ColName:  model.NewCIStr(fmt.Sprintf("join_agg_%d", len(schema))), // useless but for debug
+			FromID:   id,
+			Position: len(schema),
+			RetType:  aggFunc.GetType(),
+		})
+	}
+	aggFunc.SetArgs(expression.Schema2Exprs(schema[len(schema)-len(result):]))
+	aggFunc.SetMode(expression.FinalMode)
+	return result, schema
+}
+
+func allFirstRow(aggFuncs []expression.AggregationFunction) bool {
+	for _, fun := range aggFuncs {
+		if fun.GetName() != ast.AggFuncFirstRow {
+			return false
+		}
+	}
+	return true
+}
+
+func (a *aggPushDownSolver) produceNewAggPlan(aggFuncs []expression.AggregationFunction, gbyCols []*expression.Column, join *Join, childIdx int) (LogicalPlan, bool) {
```

Comment:
The function name is not specific enough, there are many ways to produce new aggregate plan, I can't be sure this is actually doing push down just by reading the name.
And please add comments on this function.